### ISSUE TITLE
WSL2 X11 output corruption

### ISSUE THREAD

**Issue by deviantbit** [microsoft/wslg#1150]
### Windows Version

Version	10.0.22631 Build 22631

### WSL Version

 2.0.9.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.133.1-1

### Distro Version

Ubuntu-22.04

### Other Software

Any and all applications that use xlib or X Windows.

### Repro Steps

Launch an xeyes, or emacs.

### Expected Behavior

![image](https://github.com/microsoft/WSL/assets/95936426/a6ec8e74-f293-4d86-b09b-13e391a6313a)

### Actual Behavior

![image](https://github.com/microsoft/WSL/assets/95936426/f1554e46-a4b8-485a-82a7-ee1a4ab87787)

### Diagnostic Logs

Nothing in logs to report.

**Comment by deviantbit** [comment]
I plugged in a 1080p monitor. It had the same issue.

**Comment by etxaleku** [comment]
I also noticed that whenever I try to run any GUI apps like emacs for example, my %AppData%\Local\Temp\DiagOutputDir\RdClientAutoTrace folder is filling up with hundreds of 4MB files...

![trace-files](https://github.com/microsoft/wslg/assets/1469850/aa9f6bf0-1dcd-401f-a530-8bea623798a5)

**Comment by deviantbit** [comment]
This is rasterization problem. 

![image](https://github.com/microsoft/wslg/assets/95936426/8e6d32bb-58cf-4a2f-bea7-ba0905b29a3b)

**Comment by etxaleku** [comment]
I got mine completely working!  It's rock solid and I couldn't be happier... well, I guess I could win the Lotto but...  :)

So I tried a ton of stuff but the one that got everything working was upgrading my video driver.  So throughout Microsoft's WSL/WSLg web pages they do mention that it's very important for WSLg/Wayland to work that you update your video driver to the latest from the manufacturer. But then both Microsoft (laptop vendor) and Intel (driver vendor) warn that if you update from Intel (the driver vendor) then you risk losing any special customizations/configurations that the OEM vendor (Microsoft in this case) may have made and worst case you could even lose your video entirely (the warnings look a bit ominous). They both recommend going to the laptop vendor site and upgrading from there as the safe path - so that's what I had (previously) done. But in the case of the Surface Pro Microsoft does offer a single package that upgrades all the Surface drivers and firmware and puts everything back to "out of box" so I applied that (early on) but no joy. So armed with that as a "backup" I went ahead, held my breath, and upgraded from the Intel (driver vendor) site at www.intel.com/content/www/us/en/support/intel-driver-support-assistant.html being careful to select the right video hardware and "generation", etc. The vendor driver update went smoothly (I may have passed out holding my breath once or twice during the process, it took a while) but then after a reboot I brought up emacs and it seems rock solid. No video weirdness whatsoever, new frames are fine, splitting windows is flawless, everything is exactly as I've known emacs to be for, well, a long time.  Even web browsing works (I'd never even tried that before)!

And no more trace files have shown up in my %AppData%\Local\Temp\DiagOutputDir\RdClientAutoTrace folder.

Also, all other WSL GUI apps (gnome-tweaks, octave, gnuplot, etc) are also all working smoothly now.  Octave and gnuplot integrate and work from emacs smoothly.

I cannot say enough great things about WSL/WSLg - it's the environment I've always wished for and maybe a bit more!  

Thanks to everyone that worked on and supports WSL/WSLg!

**Comment by deviantbit** [comment]
Something else resolved it. I'm not on a laptop, and using the latest drivers. I went back several generation of drivers and it still had the problem. So you did something else.

**Comment by ksylvan** [comment]
I'm on a desktop, with an external display, seeing the same thing.

![image](https://github.com/microsoft/wslg/assets/1226059/f2ecfc80-ea13-476f-b727-04ae7f38aa87)

I have an NVIDIA GeoForce RTX-4090 and the latest drivers.

**Comment by etxaleku** [comment]
Hmm, I see what you mean. Are you also seeing those trace files piling up in your %AppData%\Local\Temp\DiagOutputDir\RdClientAutoTrace folder?  

For me, I really feel like the driver update made the difference. But there’s a lot of components involved here so I can imagine that every situation has its own nuances.  I kept pretty careful notes and I went back and reviewed my command history, and I couldn’t find anything else that looked like it really had any hope.  I installed libgccjit-11-dev if that makes a difference.

**Comment by ksylvan** [comment]
> Hmm, I see what you mean. Are you also seeing those trace files piling up in your %AppData%\Local\Temp\DiagOutputDir\RdClientAutoTrace folder?

Yes @etxaleku

**Comment by keesj** [comment]
I have had similar issues (on and off) for .. about the last 2 or 3 weeks. E.g. I started seeing the issue on a Friday morning(I guess after a windows update) and was having different problems but the issue disappeared .. possibly after restarting wsl. Due to other issues over the last week (can only run WSL as administrator I install a pre-release (version 2.0.11) see https://github.com/microsoft/WSL/releases and I think the problems started to appear again.

![image](https://github.com/microsoft/wslg/assets/57112/bc9d5e87-0bb8-4641-afd6-1e970b368e0a)

I have an Intel IRISxe graphics card, dell laptop and dell usb-c docking startion/multiple displays.

**Comment by ksylvan** [comment]
I've restarted WSL (and even shutdown and restart the desktop) and can reproduce this every time, so it's a genuine WSLG bug as far as I can see.

**Comment by etxaleku** [comment]
I had a bit of a regression - so my Surface Pro 7 with WSL2 initially experienced the video "fuzzing" problem and I struggled with it for a while but eventually upgraded my Intel Iris Plus  driver from the intel web site and all my issues stopped.  No more video "fuzz" ever, all Linux graphical apps run smoothly (gnome-tweaks, etc) and no more trace files piling up in that RdClientAutoTrace folder.  Well... that is until I did some desktop screen sharing and tried to show them how it was all working... something about the desktop screen sharing caused the issue to come back - emacs and all Linux GUI apps all "fuzzed out" 100% of the time and the RdClientAutoTrace folder starting filling up with the 4MB trace files again... situation continued until I rebooted and after the reboot everything is back to rock solid working order... no more fuzz or trace files...

So it look like I have a great working environment, but I can't ever do any screen sharing and expect it to work.  :(

And probably once I do any screen sharing then I'll need to reboot to get things working again.

Oh well, it could be worse, as others have mentioned (where they don't have a "fix").

Just thought I'd document the screen sharing aspect as that might shed some insight to a WSLg developer trying to work on this issue.

**Comment by mahi1980** [comment]
I have the same issues. Updating GPU drivers didn't work (AMD Radeon Pro WX 2100, Driver 31.0.21018.6011 (2023/08/24)).
RdClientAutoTrace is being filled as well, but file size is much larger than 4MB.

**Comment by Stanzilla** [comment]
I have the same issue, also latest driver from today (nvidia)

**Comment by lucvdv-aleacsys** [comment]
Same issue here, latest Dell-provided Nvidia drivers on a Dell laptop.
I only found out about WSLg this morning after the problem occurred. I don't use GUI apps that much and thought I was still using VCXSrv, which always did the job fine. Turns out it was silently changed into using WSLg.
No wonder uninstalling (and reinstalling) VCXSrv didn't help :)

**Comment by bryghtlabs-richard** [comment]
Dell laptop with Dell Nvidia drivers, first boot of wsl2, same issue. Looks like a mismatch in stride or pixel format. Thanat0s' linked solution `apt update, apt upgrade, wsl --shutdown` fixed it for me.

**Comment by brianherold** [comment]
Thanat0s' linked solution fixed it for me as well. I'm running a Dell laptop with Dell Nvidia drivers. The issue popped up for me sometime last week after a few months of stable use.

**Comment by deviantbit** [comment]
That didn't resolve it for me. Maybe there is an update being pushed down stream that hasn't arrived.

**Comment by lucvdv-aleacsys** [comment]
I hadn't tried any GUI apps since my previous comment (2 weeks ago), so I don't know exactly when it got fixed, but the problem no longer occurs here either (see comments by brianherold and thanat0s).
I remember doing an 'apt update && apt upgrade' last wednesday, so that must have fixed it.

**Comment by deviantbit** [comment]
This has been resolved for me now.

**Comment by deviantbit** [comment]
Okay. I reopened the issue.

**Comment by etxaleku** [comment]
I agree, it is working for me (very reliably) but there are times when it regresses and the issue reappears.  The two situations where I have seen it reappear is -

R1. if I use Teams in a screen sharing session then I have seen it reappear.  Rebooting (and avoiding screen sharing) returns it to working condition.
R2. if I bring up other Windows-based apps of a "graphical" nature, like a browser, BEFORE starting WSL2 and X-based Linux apps, then the issue seems to be very reproducible - so far 100% of the time based on about 10 reiterations.

For R1 I have successfully been able to do some Teams screen sharing recently and did not have the issue reappear.  That one seems to be intermittent.  Given the frequency of the Teams screen sharing recurrence, I'm almost wondering if those few times that occurred that it might have actually been a reproduction of R2 as I wasn't aware of R2 initially. 

For R2, that one seems very reproducible.  If I reboot and the very first thing I start up is WSL2 and then emacs then everything works very reliably and I can use anything (Windows or Linux) I want for days on end without issue.  But if I reboot and bring up a browser first then when I start WSL2 and then emacs the issue appears every time.  A quick reboot returns everything to working order.

@avaranze, when you say that it works and then "always happens again after some time", how much time are you finding has to pass before it happens?  Is it hours or days?  Mine has been stable for periods on the order of 7 days or so.  The main X-based "graphical" app that I'm running is emacs though I have recently started running a Linux-based version of IntelliJ (Community Edition).

I do also see the "other" issue where X-based graphical apps are "orphaned" after a Windows sleep.  Basically if your machine goes to sleep for any reason (manual or automatic), then after resuming the graphical apps are still running according to "ps -ef" but they don't appear on the taskbar and cannot be rendered.  But this is a different issue than the one discussed here - the video "fuzzing" issue.

**Comment by keesj** [comment]
I am still having the issue

**Comment by fuzy112** [comment]
I still have this issue.  A workaround for me is to use fullscreen mode when using Emacs (`M-x toggle-frame-fullscreen`).  But other apps still don't work now.

**Comment by indigodarkwolf** [comment]
I ran into this issue on a fresh Windows 11 install. Double-checked and already had the latest drivers for my NVIDIA card.

I wound up working around it by creating a .wslgconfig file in %userprofile%, following the advice from https://github.com/microsoft/wslg/issues/1148#issuecomment-1827607993.

**Comment by jakubzeman-acc** [comment]
A simple solution that worked for me on Windows 11 was setting a default user in the Windows registry. Set the following key to your user guide (usually 1000, which is in HEX 3e8)> HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Lxss\<UUID>\DefaultUid

**Comment by tkonsta** [comment]
After reading some comments which related to the intermittent WSLg issues to running programs on windows I did some tests and for me disabling Microsoft PowerToys resulted in a reliably stable WSLg experience without UI glitches and without lagging or hanging keyboard. Maybe someone else who is running PowerToys on Windows could give this a try and confirm?

**Comment by keesj** [comment]
I have tried a few more suggestions posted here and elsewhere 
* Disabling opengl in wslconfig
```
keesj@vlsi:/mnt/c/Users/keesj$ cat .wslconfig
[wsl2]
memory=24GB

[system-distro-env]
;disable GPU in system-distro
LIBGL_ALWAYS_SOFTWARE=1
```
* Downgrading drivers

Perhaps that my setup has something else with it going wrong.
* I did upgrade the kernel / ubuntu to a recent version e.g.  23.10
* Hyper-V is disabled

Currently I am working around the issue by using xpra (but that does require me to remount /tmp/.X11-unix read/write

**Comment by Cpt-Squirrell** [comment]
I can throw my hat into the mix here. Same issue consistently as the images earlier in this thread show.
I run Intel ARC card so hey, mine might not be very surprising. For all I know though it might not be GPU related at all.

**Comment by mahi1980** [comment]
For me, updating to the latest pre-release of WSL solved the issue.
https://github.com/microsoft/WSL/releases/tag/2.1.1

**Comment by keesj-riscure** [comment]
> For me, updating to the latest pre-release of WSL solved the issue. https://github.com/microsoft/WSL/releases/tag/2.1.1

From the looks of it.  it might also have solved my problem

**Initial state**
```
WSL version: 2.0.14.0
Kernel version: 5.15.133.1-1
WSLg version: 1.0.59
MSRDC version: 1.2.4677
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.4046
```
**Upgraded**
```
wsl --update --pre-release
Checking for updates.
Updating Windows Subsystem for Linux to version: 2.1.3.
```

and so far things appear to be working. I also noticed that [2.1.1](https://github.com/microsoft/WSL/releases/tag/2.1.1)  has an upgrade of gwsl to version 1.0.60

**Comment by etxaleku** [comment]
@keesj-riscure and @mahi1980, your approach of upgrading to the WSL pre-release sounds very promising and your results sound great but one question - how stable is your environment?  Do you use it a lot and depend on it for work?  It sounds like you've both been on it for about a week, any concerns?  I use mine for work and have to depend on it to work.

One aspect I've noticed that probably significantly contributes to the seemingly random behavior is that starting from a Windows reboot, for me at least, it greatly depends on the order that I start apps.  If I reboot and start emacs or any X-application from WSL BEFORE running any Windows graphics apps then everything works and life is good.  But if I reboot Windows and bring up say a browser and pull up some web pages and then try to start any WSL X-applications then I get the video fuzzing.

Definitely planning on trying @neodon's recommendation of  ```Restart-Service -Name WSLService``` from PowerShell, that sounds promising.

**Comment by keesj-riscure** [comment]
> @keesj-riscure and @mahi1980, your approach of upgrading to the WSL pre-release sounds very promising and your results sound great but one question - how stable is your environment? Do you use it a lot and depend on it for work? It sounds like you've both been on it for about a week, any concerns? I use mine for work and have to depend on it to work.
> 

I use WSL2 / Ubuntu 99% of the time and I am quite sure my setup is unique and over the last 2 years of using WSL2 it has been a love/hate affair. I can not perform my work without it and when it works .. it works great but I do accept some breakage.

This last update resolved multiple issues for me who is forced to use a firewall/VPN. I needed to run "netsh winsock reset" after every reboot before I was able to access WSL2 and this problem is gone. A colleague of mine was able to restore X11 using https://github.com/microsoft/wslg/issues/426 (e.g. wsl --system, and then kill -9 the /usr/bin/weston ) but that did not work for me.

In the time X11 was not working I used https://github.com/Xpra-org/xpra/wiki/Download with much satisfaction. 
(I had to mount -o remount,rw /tmp/.X11-unix the X11 folder and started xpra start --start=xterm --bind-tcp=0.0.0.0:10000 )

All that to say.. that I am currently very happy with the upgrade and the relationship is currently good but dmesg is trying to tell me something different:
```
[   17.647043] WSL (2) ERROR: WaitForBootProcess:3335: /sbin/init failed to start within 10000
[   17.667864] WSL (2): Creating login session for keesj
```

> One aspect I've noticed that probably significantly contributes to the seemingly random behavior is that starting from a Windows reboot, for me at least, it greatly depends on the order that I start apps. If I reboot and start emacs or any X-application from WSL BEFORE running any Windows graphics apps then everything works and life is good. But if I reboot Windows and bring up say a browser and pull up some web pages and then try to start any WSL X-applications then I get the video fuzzing.
> 
I have had occasions where it was working but wsl --shutdown never fixed it for me.
 
> Definitely planning on trying @neodon's recommendation of `Restart-Service -Name WSLService` from PowerShell, that sounds promising.

**Comment by mahi1980** [comment]
@etxaleku I need WSL for work as well, so basically I had no real choice. Regularly trying the new pre-releases was my only hope to come back to a fully operational setup.
So far, I did not have any further issues with the pre-release.

**Comment by etxaleku** [comment]
Thanks for the feedback, gives me confidence to go to the pre-release.  

> This last update resolved multiple issues for me who is forced to use a firewall/VPN. I needed to run "netsh winsock reset" after every reboot before I was able to access WSL2 and this problem is gone.

Oh wow, I wonder if this will resolve my WSL2 issue when on VPN...  Whenever I'm on VPN I have had to use WSL1 as WSL2 would not route traffic properly.  I tried everything I could find to get it to work but nothing helped - maybe going to the pre-release of WSL2 will do it for me!

**Comment by bernhardkaindl** [comment]
In another issue, it was recommended to kill the weston of the system instance for WSDg to let it restart:
```py
# On CMD.exe or also PowerShell:
wsl --system pkill weston
```
This helped in my case now.

**Comment by etxaleku** [comment]
Thought I'd update this issue since it's still open and mention that WSL updates (and possibly Ubuntu updates?) seem to have resolved this issue for me completely!

```
WSL version: 2.4.13.0
Kernel version: 5.15.167.4-1
WSLg version: 1.0.65
MSRDC version: 1.2.5716
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26100.1-240331-1435.ge-release
Windows version: 10.0.26100.3476
```

WSL2 is very impressive and getting better all the time.
